src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/CLinker.java

```diff
@@ -54,11 +103,305 @@ private CSupport() {}
      * @throws IllegalAccessError if the runtime property {@code foreign.restricted} is not set to either
      * {@code permit}, {@code warn} or {@code debug} (the default value is set to {@code deny}).
      */
-    public static ForeignLinker getSystemLinker() {
-        Utils.checkRestrictedAccess("CSupport.getSystemLinker");
+    static CLinker getInstance() {
+        Utils.checkRestrictedAccess("CLinker.getInstance");
         return SharedUtils.getSystemLinker();
     }
 
+    /**
+     * Obtain a method handle which can be used to call a given native function.
+     *
+     * @param symbol   downcall symbol.
+     * @param type     the method type.
+     * @param function the function descriptor.
+     * @return the downcall method handle.
+     * @throws IllegalArgumentException in the case of a carrier type and memory layout mismatch.
+     */
+    @Override
+    MethodHandle downcallHandle(Addressable symbol, MethodType type, FunctionDescriptor function);
+
+    /**
+     * Allocates a native stub segment which contains executable code to upcall into a given method handle.
+     *
+     * <p>The base address of the returned stub segment can be passed to other foreign functions
+     * (as a function pointer). The returned segment is <em>not</em> thread-confined, and it only features
+     * the {@link MemorySegment#CLOSE} access mode. When the returned segment is closed,
+     * the corresponding native stub will be deallocated.</p>
+     *
+     * <p>The method type of the target method handle is used for linking</p>
+     *
+     * @param target   the target method handle.
+     * @param function the function descriptor.
+     * @return the native stub segment.
+     * @throws IllegalArgumentException in the case of a carrier type and memory layout mismatch.
+     */
+    @Override
+    MemorySegment upcallStub(MethodHandle target, FunctionDescriptor function);
+
+    /**
+     * The prefix of all ABI attribute names.
+     */
+    String ABI_ATTR_PREFIX = "abi/";
```

Comment:
does this go here? Not sure what it buys to clients.

Reply:
I added this for fixing jextract. Jextract needs to filter out these attributes before making condys, since otherwise you get an access error when resolving. So, I thought that specifying a common name prefix would be the best way to make that possible, without exposing the actual attribute names that need to be filtered.

Reply:
not super sure I get this. I imagine that jextact knows well when it needs to emit a layout corresponding to a primitive type, so, if that's the case, it should just emit an access to a static constant in Clinker and be done, rather than trying to dissect whatever intermediate layout jextract might have generated internally. I guess what I'm saying is - let's say that jextract has some layout with 4 attributes for C_INT; then its backend should just emit an access for CLinker.C_INT, and re-add the name attribute, which is really the only important part we need?

Reply:
Yes, the name and alignment are important, but maybe we'd want to add other attributes in the future? Or maybe not us, but another tool that wants to emit more attributes. It seemed fair that if we added some attributes that clients were not allowed to look at, at least we could help them with stripping if needed.

I guess another approach is to not let jextract's Type.Primitive use the CLinker layout constants, but let it generate it's own ValueLayouts, attach name and alignment attributes, and then copy them over to the condy/source using withAttribute calls when emitting classes/sources.

Reply:
I'll remove it for now, and then we can figure out what to do for jextract separately.